Switch to cmake

Wenzel opened the pull request:
I couldn't sleep and I'm already back from my future next week.
This should help fix #673 

A quick `CMake` tutorial:
~~~
mkdir build
cd build
cmake ..
make
~~~

:tada: 
![capture d ecran de 2018-10-21 04-09-39](https://user-images.githubusercontent.com/964610/47262089-3c5c0e80-d4e7-11e8-8812-1579aed67dab.png)


`CMake` pro tip:
use `ccmake ..` once you have alredy configured the build with `cmake ..`, this will call an interactive interface in your terminal to configure the build options:
![capture d ecran de 2018-10-21 04-03-17](https://user-images.githubusercontent.com/964610/47262058-4cbfb980-d4e6-11e8-98d3-7e61024df5e9.png)

# TODO

## Options
- [x] `ENABLE_XEN`
- [x] `ENABLE_KVM`
- [x] `ENABLE_FILE`
- [x] `ENABLE_SHM_SNAPSHOT`
- [x] `ENABLE_WINDOWS`
- [x] `ENABLE_LINUX`
- [x] `ENABLE_FREEBSD`
- [x] `ENABLE_XENSTORE`
- [x] `ENABLE_VMIFS`
- [x] `ENABLE_CONFIGFILE`
- [x] `ENABLE_ADDRESS_CACHE`
- [x] `ENABLE_PAGE_CACHE`
- [x] `ENABLE_STATIC`
- [x] `MAX_PAGE_CACHE_SIZE`
- [x] `REKALL_PROFILES`
- [x] `ENABLE_SAFETY_CHECKS`
- [x] `BUILD_EXAMPLES`
- [x] `debug` -> `CMAKE_BUILD_TYPE`

## Check for libraries
- [x] check for `libvmi_request.h`
- [x] check for architecture `I386` `X86_64` `ARM32` `ARM64` 
- [x] find `GLib`
- [x] find `json-c` (if rekall_profile or kvm driver)
- [x] find `check` (if tests)
- [x] find `libvirt` (if kvm)
- [x] find `dlopen` (if xen or kvm drivers)
- [x] check xen deps
  - [x] `xenctrl.h xen/io/ring.h`
  - [x] `hvmmem_access_t`
  - [x] check for `xenstore.h xs.h` (if xenstore)

# Misc
- [x] handle `flex` and `bison` to generate the config file parser, and add the generated sources to the build
- [x] manage static and dynamic build of `libvmi`
- [x] make install
- [x] use `CMake` as a test suite ?
- [x] remove autotools
  - [x] remove autotools patterns in `.gitignore`
- [x] update `debian/control` dependencies
- [x] update README and documentation
- [x] update `.travis.yml` 

The `autotools` build system will be removed at the end.

Any help and feedback is welcome !

Wenzel commented:
ping @tklengyel what about `vmifs` ? are we still maintaining this ?

tklengyel commented:
Don't forget that LibVMI also targets ARM architectures, so CMake has to detect what architecture it is going to be compiled for automatically. 

>  add option to manage vmifs (is this still a thing ??)

Yes, it is still a useful tool.

Wenzel commented:
@tklengyel i added a module to detect the architecture, but i can't test it under `ARM` to see if the `CMAKE` variable has the value i expect.

I also had to make a small modification to `core.c`:
https://github.com/libvmi/libvmi/pull/674/commits/083ccb0cba89025f23cf949bbba453c2f246e76d#diff-d9bf9c5a9edd83e259871df1d021474bR787

Wenzel commented:
I think i support all the options and detect all the required libraries by now.
Please @tklengyel and @smaresca give it a try, inspect `build/config.h` and give me a feedback if i need to change anything.

Wenzel commented:
Ok, so 2 more things:
- the `doc` target will build the documentation with `Doxygen`
- the `static_analysis` target will analyze the code with `scan-build` and provide a report (need `clang`) :wink: 
![capture d ecran de 2018-10-23 06-25-27](https://user-images.githubusercontent.com/964610/47333763-cb565b80-d68c-11e8-89db-f1cb23545d56.png)

Wenzel commented:
@tklengyel can i have some documentation about how the test suite is supposed to be ran ? and in which VM context ?
Do you have a GSoC student working on that ?

I wanted to see if i could integrate it with `CMake test` and use the `CDash` dashboard to display the results.

smaresca commented:
> Ok, so 2 more things:
> 
>     * the `doc` target will build the documentation with `Doxygen`
> 
>     * the `static_analysis` target will analyze the code with `scan-build` and provide a report (need `clang`) 😉

It's generally my preference that the clang target be marked as a todo for inclusion in our continuous integration pipeline. I want a PR to be marked as needing work if it has egregious issues.

Wenzel commented:
the latest commit compile all the tests except `test_shm_snapshot` which is based on an old version of the libvmi API, where the return value where not unified under `status_t`.

@tklengyel the autotools target was `make check`, not `make test` :)
![check](https://user-images.githubusercontent.com/964610/47440352-a4944400-d7b6-11e8-9bd5-711abe58c774.png)

I didn't looked into the details of how the tests are supposed to be run, but the target is there now:
![toto](https://user-images.githubusercontent.com/964610/47440410-c55c9980-d7b6-11e8-8f2f-203f0109c636.png)

tklengyel commented:
@Wenzel this how the test suite gets run by Jenkins (WinXP SP2 test as an example):

```
./autogen.sh
./configure --disable-kvm --disable-xen
./tests/run_tests.sh /opt/images/xp-laptop-2005-06-25.img
sudo make clean

tklengyel commented:
@Wenzel additional TODO: update `.travis.yml`

Wenzel commented:
And the test suite is **working** ! :tada: 
I updated the error message:
![screenshot_20181105_190102](https://user-images.githubusercontent.com/964610/48017201-f5e4f180-e12d-11e8-8987-6e65c1968dee.png)

Here is the test suite passing (`sudo -E make test`):
![screenshot_20181105_190324](https://user-images.githubusercontent.com/964610/48017226-04cba400-e12e-11e8-898e-aa105aea73bd.png)

And a more verbose version using `sudo -E ctest -V`
![screenshot_20181105_190310](https://user-images.githubusercontent.com/964610/48017255-16ad4700-e12e-11e8-9bd3-a05dbc8ffab4.png)
`...`
![screenshot_20181105_190257](https://user-images.githubusercontent.com/964610/48017257-19a83780-e12e-11e8-97d7-d47731406069.png)

Wenzel commented:
latest commits handles the installation of libvmi, headers and the 4 examples:
![screenshot_20181105_195756](https://user-images.githubusercontent.com/964610/48020049-3dbb4700-e135-11e8-8796-f05612ced101.png)

What about the static build, do we need take care of it ?

tklengyel commented:
> What about the static build, do we need take care of it ?

Yes, `libvmi.a` should be installed alongside the so as well unless expressly disabled during configuration (currently with `--disable-static`)

Wenzel commented:
Ok, I got the shared and static builds.
`shared` is controlled by `ENABLE_STATIC`, enabled by default.
![screenshot_20181105_211324](https://user-images.githubusercontent.com/964610/48024744-b58f6e80-e141-11e8-8d40-6cb55ad2179e.png)

This is what `autotools` is producing in `/usr/lib64`:
![screenshot_20181105_213030](https://user-images.githubusercontent.com/964610/48024890-233b9a80-e142-11e8-851a-de35ef8f303d.png)

I don't have the same output:
- `libvmi.la` (`libtool` file ??)
- `libvmi.so` being a symlink to `libvmi.so.0.0.13`, is this necessary ? :thinking:

tklengyel commented:
FYI: https://gitlab.kitware.com/cmake/community/wikis/contrib/macros/LibtoolFile

Wenzel commented:
@tklengyel do we absolutely need to provide a libtool `libvmi.la` file ?
The Macro that you linked is not adapted to the recent evolutions of CMake, and I have to use hacks to make it work....

tklengyel commented:
@Wenzel I don't think we strictly need it. @smaresca thoughts?

Wenzel commented:
I found how to set the `soname` and install the symlinks:
![screenshot_20181106_041218](https://user-images.githubusercontent.com/964610/48040706-57cb4880-e17a-11e8-8c4d-a15268ef80cc.png)

I have set the library version to `0.13.0`, as it's the actual project version at the moment. (`major`.`minor`.`patch`)
However, autotools was installing `libvmi.0.0.13`.
What is the signification of the suffix ?

Wenzel commented:
sorry for the spam, i had to debug the travis build since i have a different configuration.
I'm blocked with a linking error on `check_libvmi`
https://travis-ci.org/libvmi/libvmi/jobs/451381240#L852

I don't understand what's happening because i specified the right cflags and ldflags from `check` to the target, and it won't compile.
What is strange is that it works on my laptop (`fedora 29`) ...
Maybe this is a `CMake` error ? `trusty` is quite old now :/

Wenzel commented:
> But on Travis is there is not much point compiling the check suite, that will only be run on Jenkins.

I would argue that we still need to test if there is no compilation errors in the test suite, which is what is happening here.

Thanks for the link, I reintroduced manually linking with the `pthread`, and it works ! :tada:

Wenzel commented:
I had to disable the test suite when we are testing the build with
- cache disabled
- config file disabled
- rekall profiles disabled

because `check_libvmi` was linking against those.

Wenzel commented:
Almost all green, except
- `astyle`: can you modify your regex so that it doesn't try to check the syntax of my `.cmake` modules ? :slightly_smiling_face: 
https://travis-ci.org/libvmi/libvmi/jobs/451434935#L532
- `clang --analyze`: apparently clang found some errors :ok_hand:

tklengyel commented:
You can modify anything you need in the `.travis.yml` to adjust it to the new build system

Wenzel commented:
status: 
I'm waiting on @smaresca regarding `libtool` `libvmi.la` file, and on @tklengyel for the correct `libvmi` [suffix](https://github.com/libvmi/libvmi/pull/674#issuecomment-436117483)

smaresca commented:
checking re: libvmi.la now

Wenzel commented:
btw: apparently we have a race-condition in the autotools makefiles which can result in `grammar.h` not being generated while another job will try to use as input of compilation unit:
https://travis-ci.org/libvmi/python/jobs/454238037#L700

I had to change my travis from `make -j4` to `make`:
https://github.com/libvmi/python/pull/27

tklengyel commented:
@Wenzel yes, I believe that was reported a while ago. Not sure what would be the proper fix for it.

Wenzel commented:
I believe that CMake can deal with this automatically, since i scan the dependencies of a library and builds them recursively.
The `libconfig.so` private library i created depends on the output of `flex` and `bison` commands.
https://github.com/libvmi/libvmi/pull/674/files#diff-ba5630ac82418f33e80654086c15a20eR20

Wenzel commented:
Good news, I asked on IRC, and someone helped me to generate the libtool file, at least partially.

A few things:

## Version number
in the project, we set the version as `0.13.0` (`MAJOR`.`MINOR`.`PATCH`), which results in
cmake libtool
~~~
library_names='libvmi.so.0.13.0 libvmi.so.0 libvmi.so'
~~~

autotools libtool
~~~
library_names='libvmi.so.0.0.13 libvmi.so.0 libvmi.so'
~~~

So we need to decide how to manage the version number.

## dependencies
I had to hardcode the dependencies in the libtool file:
~~~
dependency_libs=' -lglib-2.0 -ljson-c -lm -ldl'
~~~

The rest should be fine.

Wenzel commented:
After looking at the interfaces in detail, we can drop most of the libraries that we expose
as public dependencies:
~~~
dependency_libs=' -lglib-2.0 -ljson-c -lm -ldl'
~~~
the library we need is:
- `glib-2.0` since we export a `GSList*` in `libvmi_extra.h`

the rest is used by the libvmi implementation and already linked to `libvmi.so`
What do you think guys ?

Wenzel commented:
actually, after a bit of searching on this field:
~~~
dependency_libs

    With this list, libtool aim to compensate the lack of dependency information
    in static archives; indeed this list provides a similar list to the one in DT_NEEDED tags for ELF files.
    Contrarily to the ELF dependencies, the list does not include version information of any kind, 
    as it's designed to work in turn with static archives, which do not have version information either.
    Also, these dependencies are used for both static and dynamic link, causing no small headache for 
    developers and distributors alike, as it is well possible that different versions of the same dependency   
    are brought in, due to this list. 
~~~

so it has been designed to for static libraries, i need to provide the same dependencies as the .so is using.

Wenzel commented:
Okay, problem solved ! :tada: 

I just need your comments on the version string, and then we can make a global review of this PR :)

Wenzel commented:
Ping @tklengyel and @smaresca, waiting for your review.
Do you think we can merge this before christmas ?

tklengyel commented:
@Wenzel Sorry, I was under the impression that there are still TODO items on the list before this is in a mergaable state. At the least I would like to see it pass the Travis checks. Once that's done I will give it a spin and if it's all good adjust Jenkins. Once Jenkins passes the PR as well we can merge it.

Wenzel commented:
@tklengyel i fixed the build, after the last refactoring, i thought everything was okay, but i had to find a workaround on a CMake bug.

One of the jenkins build fails because it still use autotools, and as for Travis, it's not triggered anymore but my latest push ??? something happened ?

Wenzel commented:
merge conflicts, I'm on it.

Wenzel commented:
Status:
- Jenkins: `PR_LibVMI_Xen_Test_Ubuntu1204-64`: 
~~~
+ ./autogen.sh
aclocal: error: 'configure.ac' is required
~~~


- Travis: `clang --analyze`: 3 bugs found

tklengyel commented:
@Wenzel Don't worry about Jenkins right now, it will need to be manually adjusted to use cmake after this PR is in a good state.

tklengyel commented:
>     Travis: clang --analyze: 3 bugs found

Please fix these as well - even if they are not introduced in this PR. Perhaps in a separate PR? Thanks!

Wenzel commented:
The fixes are not entirely obvious to me:
![screenshot_20181220_165049](https://user-images.githubusercontent.com/964610/50295038-7a3dca80-0477-11e9-8871-50108b2f0568.png)
I'd rather you to take a look at them.

tklengyel commented:
Hm, well, for 1&3 it looks like bugs in the shm code, which I think is something we should deprecate now completely. 2 is something to ignore as its in the flex generated config parser code. 4 is also an API that should be deprecated.

Wenzel commented on tests/CMakeLists.txt:
```diff
@@ -0,0 +1,63 @@
+
+# doesn't compile with latest changes in libvmi API
+# add_library(test_shm_snapshot STATIC test_shm_snapshot.c)
+# target_link_libraries(test_shm_snapshot vmi_shared ${Check_LIBRARIES})
```
reminder: @tklengyel don't forget to update this test later

Wenzel commented on libvmi/driver/xen/libxs_wrapper.h:
```diff
@@ -23,17 +23,17 @@
 struct xen_instance;
 
-#ifndef HAVE_LIBXENSTORE
+#if (HAVE_LIBXENSTORE == 0)
```
@tklengyel not sure if this is the best syntax to test a C preprocessor define, what do you think ?

tklengyel replied:
I don't think you need paranthesis

Wenzel commented on libvmi/core.c:
```diff
@@ -784,11 +784,11 @@ vmi_get_library_arch()
 #ifdef I386
     return VMI_ARCH_X86;
-#elif X86_64
+#elif defined(X86_64)
```
I had to modify the libvmi code here, make sure it's OK for you.

tklengyel commented:
@Wenzel please make the changes necessary to make Travis pass.

Wenzel commented:
@tklengyel do you mean disabling the static analysis test ?
because we will always get at least a bug in flex/bison generated code, and the rest of the bugs are located in code that should be deprecated, as you said.

Wenzel commented on CMakeLists.txt:
```diff
@@ -0,0 +1,86 @@
+#-----------------------------------------------------------------------------
+#                           GENERAL CONFIGURATION
+#-----------------------------------------------------------------------------
+project(LibVMI VERSION 0.13.0 LANGUAGES C)
```
Another hesitation I have is regarding the project version.
The library produces by CMake will be `libvmi.so.0.13.0`
But we want `libvmi.so.0.0.13`
Should we change the project version to match the library name we want to get ?

tklengyel commented:
> @tklengyel do you mean disabling the static analysis test ?

No.

> because we will always get at least a bug in flex/bison generated code 

After checking the travis script, the config code is already skipped when doing clang --analyze. You should check why that skip no longer works

> the rest of the bugs are located in code that should be deprecated, as you said.

So go ahead and deprecate them.

Wenzel commented:
Wait, what do you mean by deprecating these APIs ? :thinking: 
should we remove `walkthrough_shm_snapshot_pagetable` and everything that depend on it ?

This PR is already doing a lot of changes, so i would rather work on an API deprecation in a different PR.

tklengyel commented:
@Wenzel sure, deprecating those APIs in a separate PR is fine

Wenzel commented:
@tklengyel I forced the travis test `static_analysis_test` to pass, so Travis is green.
We will remove this in the API deprecation PR.

I'm waiting on your review of the library version issue that I pointed out earlier, and your general review of the PR.

Wenzel commented:
Then let's deprecate those API first.
I would prefer you handle this because it's your code, so you know better what should be removed, and how.

I'm not confident doing those changes, and will ask you confirmation multiple times before i can even submit a PR in my opinion :)

And btw congrats for the `bareflank` driver :+1: 
Do we have some news about the `VirtualBox` one ?

smaresca commented:
@Wenzel I am hopeful that I'll be able to share information about the vbox driver in January. Some work is required to rebase against the current head, but it's currently usable as-is. 

This cmake PR will actually help remove some of the unpleasant bits in the vbox driver which were imposed by the VirtualBox build system at the time (also cmake).

tklengyel commented:
@Wenzel unfortunately it is unlikely I will have the time to do those required changes in the near future

Wenzel commented:
still one bug left in `step_and_reg_events`, waiting for #699 , and will rebase.

tklengyel commented:
@Wenzel I would also like to ask you to wait for #693 to be merged and rebase on that as it adds assembly source files as well. I'm not up-to-speed on cmake to easily make that switch right now.

Wenzel commented:
weird that a static analysis error is raised here with the old `clang-5` from Trusty.
the warning is not raised in https://github.com/libvmi/libvmi/pull/708

Wenzel commented:
And Travis is green here too @tklengyel ;)
next steps:
- adapt jenkins tests
- review libvmi version

tklengyel commented:
@Wenzel I started the update process on the jenkins test rig, it is going to take some time as the machine needs some additional attention too - upgrading to stretch, upgrading Xen, etc..

tklengyel commented:
@libvmi-jenkins Retest this please

tklengyel commented:
I'm getting the error `make: *** No rule to make target 'check'.  Stop.` on Jenkins

tklengyel commented:
@Wenzel could you make sure that running https://github.com/KVM-VMI/libvmi/blob/switch_to_cmake/tests/run_tests.sh is something that works as-is in this PR?

tklengyel commented:
@Wenzel please also rebase this PR on the latest master. When it's being merged I'm getting some strange failures when I run `ctest -v` via Jenkins:

```
# ctest -V
UpdateCTestConfiguration  from :/var/lib/jenkins/workspace/PR_LibVMI_Xen_Test_Winxpsp2/DartConfiguration.tcl
UpdateCTestConfiguration  from :/var/lib/jenkins/workspace/PR_LibVMI_Xen_Test_Winxpsp2/DartConfiguration.tcl
Test project /var/lib/jenkins/workspace/PR_LibVMI_Xen_Test_Winxpsp2
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: test_libvmi

1: Test command: /var/lib/jenkins/workspace/PR_LibVMI_Xen_Test_Winxpsp2/tests/check_libvmi
1: Test timeout computed to be: 10000000
1: Running suite(s): LibVMI
1: error: syntax error
1: VMI_ERROR: Failed to read config file.
1:  Windows XP	GUID: bd8f451f3e754ed8a34b50560ceb08e31
1: 	Kernel filename: ntkrnlpa.pdb
1: 96%: Checks: 27, Failures: 1, Errors: 0
1: /var/lib/jenkins/workspace/PR_LibVMI_Xen_Test_Winxpsp2/tests/test_init.c:198:P:LibVMI Init:test_libvmi_init1:0: Passed
1: /var/lib/jenkins/workspace/PR_LibVMI_Xen_Test_Winxpsp2/tests/test_init.c:184:P:LibVMI Init:test_libvmi_init2:0: Passed
1: /var/lib/jenkins/workspace/PR_LibVMI_Xen_Test_Winxpsp2/tests/test_init.c:165:F:LibVMI Init:test_libvmi_init3:0: vmi_init_complete failed
```

When I just take your branch and rebase it on master it seems to work so I don't know what's that about.

Wenzel commented:
> I'm getting the error make: *** No rule to make target 'check'. Stop. on Jenkins

`check` is the test target with `autotools`.
On `CMake` it's just `test`
so `make test`.

Can you change jenkins to use `make test` instead ?

Wenzel commented:
I rebased on `libvmi:master`.

If you run `ctest --verbose` in the `build` drectory you will see the following message:
~~~
UpdateCTestConfiguration  from :/home/wenzel/Projets/libvmi/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/wenzel/Projets/libvmi/build/DartConfiguration.tcl
Test project /home/wenzel/Projets/libvmi/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: test_libvmi

1: Test command: /home/wenzel/Projets/libvmi/tests/check_libvmi
1: Test timeout computed to be: 10000000
1: !! Please specify a VM in LIBVMI_CHECK_TESTVM env var
1/1 Test #1: test_libvmi ......................***Failed    0.00 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.00 sec

The following tests FAILED:
          1 - test_libvmi (Failed)
Errors while running CTest
~~~

`1: !! Please specify a VM in LIBVMI_CHECK_TESTVM env var`

Specify the variable with `export LIBVMI_CHECK_TESTVM=ubuntu_x64`, as you would have done it in `run_tests.sh`, and run your tests:

`sudo -E ctest --verbose`
or
`sudo -E make test`

tklengyel commented:
> Can you change jenkins to use make test instead ?

Jenkins runs the script in the tests folder. So if it changes you need to change it there. For now I changed it to just run ctest -V and you can see that it fails at http://libvmitest.zentific.com:8080/job/PR_LibVMI_Xen_Test_Winxpsp2/1588/console

tklengyel commented:
I can't figure out why that one test fails when it is running through Jenkins. When I run it manually it works. If you can figure out please push a fix. If not, at this point I'm OK with just removing that test case altogether from the test suite.

Wenzel commented:
quickly ran some tests but i will run out of battery soon.
Here Windows XP, the vmi_init tests are working, but the PE parser fails for some reason:
![screenshot_20190203_163849](https://user-images.githubusercontent.com/964610/52178719-43770600-27d2-11e9-9502-bb16363535cc.png)

Wenzel commented:
If i remove the peparsing test, the target works for me, here you have make test, vmi-process-list and vmi-module-list :ok_hand: 
![screenshot_20190203_165228](https://user-images.githubusercontent.com/964610/52178916-2f340880-27d4-11e9-9f69-beee5e06d707.png)

tklengyel commented:
That's odd that we have tests now failing inexplicably. Anyway, if you can comment out `test_libvmi_init3` for now in the test suite I could finish up moving the rest of Jenkins over to cmake and finally merge this PR.

Wenzel commented:
> I could finish up moving the rest of Jenkins over to cmake and finally merge this PR.

hold up !

We have to check the MAJOR.MINOR.PATCH libvmi version first:
https://github.com/libvmi/libvmi/pull/674#issuecomment-436117483

right now:
├── grammar.h
├── libtool_template.gen
├── libvmi.a
├── libvmi.la
├── libvmi.so -> libvmi.so.0
├── libvmi.so.0 -> libvmi.so.0.13.0
├── libvmi.so.0.13.0
├── Makefile

Wenzel commented:
Also, the `Changelog` file is empty, but nonetheless in the git.
Was it required by `autotools` (also NEWS and AUTHORS files ?)
AUTHORS could be integrated in the README.

Wenzel commented:
And `Jenkins` is green !
Now `Travis` is red...
Could you rerun https://travis-ci.org/libvmi/libvmi/jobs/488188349 please ? :thinking:

Wenzel commented:
@tklengyel i fixed the libvmi version number:
~~~CMake
project(LibVMI VERSION 0.0.13 LANGUAGES C ASM)
~~~

~~~
├── grammar.h
├── libtool_template.gen
├── libvmi.a
├── libvmi.la
├── libvmi.so -> libvmi.so.0
├── libvmi.so.0 -> libvmi.so.0.0.13
├── libvmi.so.0.0.13
├── Makefile
~~~

Wenzel commented:
I removed my last commit and forced the SONAME to be 0, and the VERSION to be 0.MAJOR.MINOR.
I hope this is what autotools was doing.
~~~cmake
# set soname
set_target_properties(vmi_shared PROPERTIES
    VERSION
    0.${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}
    SOVERSION 0
)
~~~

Wenzel commented:
looks like we have a jenkins job using `autogen.sh`
`/tmp/jenkins6693902203384129098.sh: ./autogen.sh: not found`

tklengyel commented:
Yes, there were other PRs I wanted to get tested today. I'll move the whole of Jenkins over to cmake over the weekend.

Wenzel commented:
> So this is not correct. This should get installed when Windows is enabled.

the last commit was confusing because i refatored the code while rebasing.
I moved `vmi-win-guid` and `vmi-win-offsets` in another `install` directive, if Windows is enabled.

> There is also another new example that is Windows specific that is missing from here

I only see `win-guid.c` and `win-offsets.c` in examples ?

tklengyel commented:
Yeap, it's just vmi-win-offsets.

Wenzel commented:
I think it's all good:
https://github.com/libvmi/libvmi/pull/674/files#diff-2a223b2dbe4b8adba24dcfb102869d01R54

tklengyel commented:
@libvmi-jenkins Retest this please

tklengyel commented:
So unfortunately yet another test is now failing that has worked previously:
http://libvmitest.zentific.com:8080/job/PR_LibVMI_Xen_Test_Windows10/317/console

tklengyel commented:
@libvmi-jenkins Retest this please

tklengyel commented:
It turns out the Windows 10 test was actually buggy to begin with so I just disabled it for now. The ARM cross-compile test is something that would be great if you could update - and in fact, perhaps add that info to the README as well as it is not straight forward how to do that with cmake. Thanks!

Wenzel commented:
Cross compiling is more complicated than simply setting a CMake variable on the command line.
Apparently you need to write a toolchain file specifying your build environement and which compilers should be used:
https://gist.github.com/farleylai/c329dba2996ef7acde1b

Not sure who will use that feature ?
In which context do you need to generate and ARM libvmi library on an x86 host ?

tklengyel commented:
We don't have any ARM boards to do testing on so the only way to ensure LibVMI doesn't break for ARM is by doing cross-compiling and testing it on a memory dump gathered from an ARM VM. Currently both the cross-compile test and file-based tests are broken when used with cmake. So that's a big step back from the current CI setup.

tklengyel commented:
Also consider the fact that a lot of ARM boards are low-power devices which are not really suited for setting up a complete development environment on to do native compilation of tools. For those cases it makes a lot of sense to cross-compile on x86 and only deploy the binaries on the target boards.

tklengyel commented:
Anyway, I'll merge this as-is and figure out something later for the cross-compile tests.
